hashicorp/waypoint
# Fix bring-your-own-alb for plugin/ecs

Fixes https://github.com/hashicorp/waypoint/issues/4456
Replaces https://github.com/hashicorp/waypoint/pull/4457

### What changed

- Introduces a new field `load_balancer_arn` to the ecs plugin's deploy config. If specified, waypoint will use this alb rather than creating it's own.
- Removes `listener_arn` field from same config

### Motivation

Prior to this, to use an existing alb (and prevent waypoint from creating a new one), you would specify a `listener_arn` in the ecs deployment config. In practice, that was broken (see the above issue)

This approach wan't ideal anyway. Waypoint needs to modify the listener on every deployment, so if someone were to create the listener through terraform and wire the output into waypoint, that terraform would also try to change the default listener on the rule away from waypoint's deployment on every run.

Waypoint should own the listener, but can inherit an existing ALB.

### Backwards-compat

While removing a field is generally a breaking change, because the field was broken anyway (see issue), I can't see how anyone could have been depending on it.

### How do I test this?

I'd recommend starting with https://github.com/hashicorp/waypoint-examples/tree/main/aws/aws-ecs/nodejs. Run a few `waypoint deploy`'s  while observing the aws alb, and notice how the default rule moves traffic to the most recent target group on every release.

To test bring-your-own-alb, you can copy the arn and subnets of the alb waypoint created in the above testing into your deploy config

Example:

```
...
  deploy {
    use "aws-ecs" {
...
      alb {
        load_balancer_arn = "arn:aws:elasticloadbalancing:us-east-1:863953081260:loadbalancer/app/waypoint-ecs-hello-izaak-2/9d50bd2cdea67345"
        subnets      = ["subnet-0bd08267365065f8e","subnet-04ff6753f3ba35e49","subnet-02c9aaa7732de7eb7","subnet-047d5841370712b24","subnet-086d533750a535ea5","subnet-0f8dd2e1826266538"]
      }
    }
  }
```

Run a few more releases, and observe the same successful behavior.

### Other fixes in this PR

- Can now manage an ALB through terraform, and have waypoint use it
- Can now allow waypoint to use an ALB with other listeners configured, without waypoint modifying the wrong listener.
